https://bugs.gentoo.org/show_bug.cgi?id=281968
http://bugs.python.org/issue6731

--- setup.py
+++ setup.py
@@ -29,6 +29,8 @@
 else:
     disable_ssl = 0
 
+exit_status = 0
+
 def add_dir_to_list(dirlist, dir):
     """Add the directory 'dir' to the list 'dirlist' (at the front) if
     1) 'dir' is not already in 'dirlist'
@@ -257,6 +259,8 @@
             print()
 
         if self.failed:
+            global exit_status
+            exit_status = 1
             failed = self.failed[:]
             print()
             print("Failed to build these modules:")
@@ -1842,6 +1846,7 @@
           scripts = ["Tools/scripts/pydoc3", "Tools/scripts/idle3",
                      "Tools/scripts/2to3"]
         )
+    sys.exit(exit_status)
 
 # --install-platlib
 if __name__ == '__main__':
